### PR TITLE
As a guest, I can register

### PR DESCRIPTION
**Pull request by hoangmirs**
Solve https://github.com/hoangmirs/go-scraper/issues/2

## What happened 👀

- Create a registration page
- Testing will be implemented on https://github.com/hoangmirs/go-scraper/pull/26

## Insight 📝
- Add registration feature, show flash messages when failed or succeeded
- Add `/register` route
- Create `BaseController` to add some common works that we re-use in several controllers
- Add some helpers to encrypt password, set some common controller attributes, convert string
- Split the tailwind's components into multiple files to resolve the slow-building bug when using Tailwind 2 + Webpack. Will merge them again when Tailwind fixed the [issue](https://github.com/tailwindlabs/tailwindcss/issues/2820)
- Add `copy-webpack-plugin` to copy the static assets to `static` folder

## Proof Of Work 📹

Create successfully

Create failed

### CODE REVIEW

File: views/shared/header.html [hunk 1]
```diff
@@ -0,0 +1,132 @@
+<header>
+  <nav class="bg-gray-800">
```
Comment (olivierobert) [inline views/shared/header.html:2]: Not sure about your long term plans as you could just be prototyping here. But prefer using class names and moving those helpers to components.

File: conf/webpack.common.js [hunk 2]
```diff
@@ -11,6 +12,14 @@ module.exports = {
   plugins: [
     // new CleanWebpackPlugin(['dist/*']) for < v2 versions of CleanWebpackPlugin
     new CleanWebpackPlugin(),
+    new CopyPlugin({
+      patterns: [
+        {
+          from: path.resolve(__dirname, "../assets/images"),
+          to: path.resolve(__dirname, "../static/images"),
```
Comment (olivierobert) [inline conf/webpack.common.js:19]: Why is this needed?

**Review by olivierobert — COMMENTED** [review]
Early feedback

File: assets/stylesheets/layouts/authentication.scss [hunk 3]
```diff
@@ -0,0 +1,9 @@
+.layout-authentication {
+  @apply bg-gray-100;
+
+  .app-main {
+    @apply max-w-7xl mx-auto py-6 sm:px-6 lg:px-8;
+  }
+
+
```
Comment (olivierobert) [inline assets/stylesheets/layouts/authentication.scss:8]: Extra lines not required

File: assets/stylesheets/app-components.scss [hunk 4]
```diff
@@ -0,0 +1,5 @@
+@import 'tailwindcss/components';
```
Comment (olivierobert) [inline assets/stylesheets/app-components.scss:1]: Not sure about the naming and structure for all 🤔 Why si it better to have `app-components.scss`, `app-base.scss` and `app-utlities.scss` instead of `application.scss`? Especially, now there is even `compoents/index.scss` file to import all components at once. It seems rather more complicated than having one file 💦

File: bootstrap/db.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	_ "github.com/lib/pq" // Postgres driver
+)
+
+func SetUpDB() {
+	err := orm.RegisterDriver("postgres", orm.DRPostgres)
+	if err != nil {
+		logs.Critical(fmt.Sprintf("Failed to register the driver %v", err))
+	}
+
+	err = orm.RegisterDataBase("default", "postgres", web.AppConfig.DefaultString("dbUrl", ""))
```
Comment (olivierobert) [inline bootstrap/db.go:19]: IS it normal that `DefaultString` is set to ""?

File: helpers/to_snake_case.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package helpers
```
Comment (olivierobert) [inline helpers/to_snake_case.go:1]: Not sure about the file naming as it seems inconsistent with the other generic name for `controller.go` and `password.go` helpers. Maybe `text.go` helpers would be more generic 💭

File: controllers/registration.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+	web.ReadFromRequest(&c.Controller)
+
+	c.Layout = "layouts/authentication.html"
+	c.TplName = "registration/new.html"
+
+	c.Data["Title"] = "Create a new account"
+}
+
+func (c *Registration) Post() {
+	userForm := forms.RegistrationForm{}
```
Comment (olivierobert) [inline controllers/registration.go:26]: A good call to externalize this business logic to another module. Let's keep controllers thin. But the variable name is a bit confusion, shouldn't it be called `registrationForm`?

```suggestion
	registrationForm := forms.RegistrationForm{}
```

File: controllers/registration.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	flash := web.NewFlash()
+
+	err := c.ParseForm(&userForm)
+	if err != nil {
+		logs.Error("Can not parse the form", err)
```
Comment (olivierobert) [inline controllers/registration.go:31]: This error is quite too generic. In which case, would this fail?

**Review by olivierobert — COMMENTED** [review]
Some early feedback again

File: assets/stylesheets/app-components.scss (hunk 4 above)
Comment (hoangmirs) [reply assets/stylesheets/app-components.scss:1]: There is an issue with Tailwind when using Webpack. Before, I set all of them into 1 file, but it took about 16 seconds to rebuild when I made the changes, but when splitting them, it can fix that issue, so I change to use multiple files, let me update the description later
https://github.com/tailwindlabs/tailwindcss/issues/2820#issuecomment-739837017

File: bootstrap/db.go (hunk 5 above)
Comment (hoangmirs) [reply bootstrap/db.go:19]: No, that function requires 2 parameters, so I need to input it.
They also provide `web.AppConfig.String` method, but it returns the value and error value, I want to skip the error checking, then went with `DefaultString` func

File: conf/webpack.common.js (hunk 2 above)
Comment (hoangmirs) [reply conf/webpack.common.js:19]: We need to set the static path for the `assets/images` folder if we need to use it. But I want to put all static files in `static` folder, then I go with this, is it okay?

File: controllers/registration.go (hunk 7 above)
Comment (hoangmirs) [reply controllers/registration.go:26]: Yes, it should

File: controllers/registration.go (hunk 8 above)
Comment (hoangmirs) [reply controllers/registration.go:31]: I found the case when we don't pass the struct pointer, it would raise the error, but because that function returns an error, then I create this block condition to check, should we ignore this? I don't want to check all the errors 🤯

File: helpers/to_snake_case.go (hunk 6 above)
Comment (hoangmirs) [reply helpers/to_snake_case.go:1]: Yeah, that's a better name

File: assets/stylesheets/screens/index.scss [hunk 9]
```diff
@@ -0,0 +1 @@
+@import "./registration";
```
Comment (Lahphim) [inline assets/stylesheets/screens/index.scss:1]: @hoangmirs Should we go with `'` or `"`? 🤔

File: controllers/base.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
+package controllers
+
+import (
+	"github.com/beego/beego/v2/server/web"
+	"github.com/hoangmirs/go-scraper/helpers"
```
Comment (gutakk) [inline controllers/base.go:5]: how about grouping import to follow our [convention](https://nimblehq.co/compass/development/code-conventions/golang/#imports)

File: controllers/base.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+// Prepare implements Prepare method for baseController
+func (c *baseController) Prepare() {
+	// Setting properties
+	helpers.SetControllerAttributes(&c.Controller)
+
+	c.LayoutSections = make(map[string]string)
+	c.LayoutSections["FlashMessage"] = "shared/flash_message.html"
+
+	if app, ok := c.AppController.(NestPreparer); ok {
```
Comment (gutakk) [inline controllers/base.go:25]:
```suggestion
app, ok := c.AppController.(NestPreparer)
if ok {...}
```
According to our [convention](https://nimblehq.co/compass/development/code-conventions/golang/#errors)

File: controllers/base.go [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package controllers
+
+import (
+	"github.com/beego/beego/v2/server/web"
+	"github.com/hoangmirs/go-scraper/helpers"
+)
+
+type NestPreparer interface {
+	NestPrepare()
```
Comment (gutakk) [inline controllers/base.go:9]: Could you please explain what is the `NestPreparer` and `NestPrepare`?

File: controllers/registration.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	baseController
+}
+
+func (c *Registration) Get() {
+	web.ReadFromRequest(&c.Controller)
+
+	c.Layout = "layouts/authentication.html"
+	c.TplName = "registration/new.html"
+
+	c.Data["Title"] = "Create a new account"
```
Comment (gutakk) [inline controllers/registration.go:22]: How about move the `Create a new account` to const variable as I saw you use this more than one places and also the `Layout` and `TplName`

File: forms/registration.go [hunk 14]
```diff
@@ -0,0 +1,55 @@
+package forms
+
+import (
+	"github.com/beego/beego/v2/client/orm"
+	"github.com/beego/beego/v2/core/logs"
+	"github.com/beego/beego/v2/core/validation"
+	"github.com/hoangmirs/go-scraper/helpers"
+	"github.com/hoangmirs/go-scraper/models"
```
Comment (gutakk) [inline forms/registration.go:8]: How about group the import?

File: forms/registration.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+type RegistrationForm struct {
+	Email                string `form:"email" valid:"Required; Email; MaxSize(100)"`
+	Password             string `form:"password" valid:"Required; MinSize(6)"`
+	PasswordConfirmation string `form:"password_confirmation" valid:"Required; MinSize(6)"`
+}
+
+func (registrationForm *RegistrationForm) Valid(v *validation.Validation) {
+	if registrationForm.Password != registrationForm.PasswordConfirmation {
+		// Set error messages of Name by SetError and HasErrors will return true
+		err := v.SetError("Password", "Password confirmation does not match")
```
Comment (gutakk) [inline forms/registration.go:20]: How about this error message?
```
Password does not match password confirmation
```

It a little bit confuse what password confirmation does not match with.

File: forms/registration.go [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+		}
+	}
+}
+
+func (registrationForm *RegistrationForm) CreateUser() error {
+	valid := validation.Validation{}
+
+	success, err := valid.Valid(registrationForm)
+	if err != nil {
+		logs.Error("Validation error:", err)
```
Comment (gutakk) [inline forms/registration.go:32]: Not sure if this what you expect. When I tried to register with passwords don't match, this printed out to the console
```
SetError error: Password confirmation does not match
```
Nothing about `Validation error:`

File: main.go [hunk 17]
```diff
@@ -2,7 +2,8 @@ package main
 
 import (
 	"github.com/beego/beego/v2/server/web"
-	_ "github.com/hoangmirs/go-scraper/routers"
+
+	_ "github.com/hoangmirs/go-scraper/bootstrap"
```
Comment (gutakk) [inline main.go:6]: How about group the import?

File: views/shared/header.html [hunk 18]
```diff
@@ -0,0 +1,59 @@
+<header class="header">
+  <nav class="navigation-pane">
+    <div class="navigation-pane__main">
+      <div class="navigation-menu">
+        <div class="navigation-menu__left">
+          <a class="navigation-menu__logo" href="/">
+            <img src="/static/images/logo.svg" alt="Go Scraper" />
```
Comment (gutakk) [inline views/shared/header.html:7]: How about use background-image for this. You can find more information [here](https://github.com/gutakk/go-google-scraper/issues/41). cc @olivierobert

**Review by gutakk — COMMENTED** [review]
Please add the minimum node version can be used in this project as I cannot use version 10 with  this project.

**Comment by gutakk** [comment]
Screen Shot 2564-01-19 at 15 52 18
This error message is not user friendly after I tried to register with duplicate email.

File: assets/stylesheets/screens/index.scss (hunk 9 above)
Comment (hoangmirs) [reply assets/stylesheets/screens/index.scss]: I fixed it to use `'`, thanks for pointing it out

File: controllers/base.go (hunk 12 above)
Comment (hoangmirs) [reply controllers/base.go:12]: I'm not sure about it, you can check it [here](https://beego.me/docs/mvc/controller/controller.md#custom-logic)
but it looks like we define an interface and the child controller can inherit from it.
I used this to define something like `before_action` in Rails

File: controllers/registration.go (hunk 13 above)
Comment (hoangmirs) [reply controllers/registration.go]: I moved them into a new function

File: forms/registration.go (hunk 15 above)
Comment (hoangmirs) [reply forms/registration.go]: I changed. Actually, we can not say `Password confirmation does not match the email` or something else, but to avoid your confusion, it should be changed

File: forms/registration.go (hunk 16 above)
Comment (hoangmirs) [reply forms/registration.go]: This just logs the error when `Valid()` doesn't work as expected, I don't know when it will raise that error, but the `Valid()` function returns a success value and an error object, so we have to handle the error

File: views/shared/header.html (hunk 18 above)
Comment (hoangmirs) [reply views/shared/header.html]: I changed to use the `background-image`, will research about the SVG inline later

File: assets/stylesheets/layouts/_authentication.scss [hunk 19]
```diff
@@ -0,0 +1,9 @@
+.layout-authentication {
+  @apply bg-gray-100;
+
+  .app-main {
+    @apply max-w-7xl mx-auto py-6 sm:px-6 lg:px-8;
+  }
+
```
Comment (Lahphim) [inline assets/stylesheets/layouts/_authentication.scss:7]: @hoangmirs We should remove some extra spaces here out. ✌️

**Review by Lahphim — APPROVED** [review]

**Review by gutakk — APPROVED** [review]

File: assets/stylesheets/app-components.scss (hunk 4 above)
Comment (olivierobert) [reply assets/stylesheets/app-components.scss:1]: Noted. Thanks for the infor

File: conf/webpack.common.js (hunk 2 above)
Comment (olivierobert) [reply conf/webpack.common.js:19]: I am not sure of the motivation to move images to `static/` 🤔 Do you mean that once the build is completed, the compiled JS and CSS would go to `/static` to you prefer to have images there too?

File: controllers/registration.go (hunk 8 above)
Comment (olivierobert) [reply controllers/registration.go:29]: This error check seems superfluous as sending the param is not from the outside but from within the method Post. So there is no risk that a pointer won't be provided 🙃

File: assets/stylesheets/components/_header.scss [hunk 20]
```diff
@@ -0,0 +1,87 @@
+.header {
```
Comment (olivierobert) [inline assets/stylesheets/components/_header.scss:1]: This violates [a few Compass rules](https://nimblehq.co/compass/development/code-conventions/css/#class-names): 
- Class names must be composed of at least two words to make each element class name specific and more descriptive.
- Do NOT use HTML tag name as class names.
Exception can be fine like `.btn`, but be careful

File: conf/webpack.common.js (hunk 2 above)
Comment (hoangmirs) [reply conf/webpack.common.js]: No, before I couldn't use assets from the `assets` folder, I have to [set the static path](https://beego.me/docs/mvc/view/static.md) to use it, then I went with putting all the static files in the same folder.
But I changed to use `assets` folder in https://github.com/hoangmirs/go-scraper/pull/23/commits/45a4a02af51b02cae2bbef6ee90a7c0dbf9fbc4b

File: assets/stylesheets/components/_header.scss (hunk 20 above)
Comment (hoangmirs) [reply assets/stylesheets/components/_header.scss]: Fixed in https://github.com/hoangmirs/go-scraper/pull/23/commits/2b9d7e9edb724629cf2ea19efd1549a92edf7cbb

**Review by olivierobert — APPROVED** [review]
